nilearn/surface/surface.py

```diff
@@ -0,0 +1,644 @@
+    sample_locations_voxel_space = (mesh_voxel_space[:, np.newaxis, :] +
+                                    offsets_voxel_space[np.newaxis, :])
+    return sample_locations_voxel_space
+
+
+def _line_sample_locations(
+        mesh, affine, segment_half_width=3., n_points=10):
+    """Locations to draw samples from to project volume data onto a mesh.
+
+    For each mesh vertex, the locations of `n_points` points evenly spread in a
+    segment of the normal to the vertex are returned. The line segment has
+    length 2 * `segment_half_width` and is centered at the vertex.
+
+    Parameters
+    ----------
+    mesh : pair of numpy.ndarray.
+        mesh[0] contains the 3d coordinates of the vertices
+        (shape n_vertices, 3)
+        mesh[1] contains, for each triangle, the indices into mesh[0] of its
+        vertices (shape n_triangles, 3)
+
+    affine : numpy.ndarray of shape (4, 4)
+        affine transformation from image voxels to the vertices' coordinate
+        space.
+
+    segment_half_width : float, optional (default=3.)
+        size in mm of the neighbourhood around each vertex in which to draw
+        samples
+
+    n_points : int, optional (default=10)
+        number of samples to draw for each vertex.
+
+    Returns
+    -------
+    numpy array, shape (n_vertices, n_points, 3)
+        The locations, in voxel space, from which to draw samples.
+        First dimension iterates over mesh vertices, second dimension iterates
+        over the sample points associated to a vertex, third dimension is x, y,
+        z in voxel space.
+
+    """
+    vertices, faces = mesh
+    normals = _vertex_outer_normals(mesh)
+    offsets = np.linspace(-segment_half_width, segment_half_width, n_points)
+    sample_locations = vertices[
+        np.newaxis, :, :] + normals * offsets[:, np.newaxis, np.newaxis]
+    sample_locations = np.rollaxis(sample_locations, 1)
+    sample_locations_voxel_space = np.asarray(
+        resampling.coord_transform(
+            *np.vstack(sample_locations).T,
+            affine=np.linalg.inv(affine))).T.reshape(sample_locations.shape)
+    return sample_locations_voxel_space
+
+
+def _sample_locations(mesh, affine, radius, kind='line', n_points=None):
+    """Get either ball or line sample locations."""
+    projectors = {
+        'line': _line_sample_locations,
+        'ball': _ball_sample_locations
+    }
+    if kind not in projectors:
+        raise ValueError(
+            '"kind" must be one of {}'.format(tuple(projectors.keys())))
+    projector = projectors[kind]
+    # let the projector choose the default for n_points
+    # (for example a ball probably needs more than a line)
+    loc_kwargs = ({} if n_points is None else {'n_points': n_points})
+    sample_locations = projector(
+        mesh, affine, radius, **loc_kwargs)
+    return sample_locations
+
+
+def _masked_indices(sample_locations, img_shape, mask=None):
+    """Get the indices of sample points which should be ignored.
```

Comment:
I have the impression that you are doing a lot of "not" operations when you are building this mask and using it. I wonder if it wouldn't be beneficial to invert the function and return the selected indices.

Reply:
I'll try to change it and see if it makes a difference. I did it like this because it was easier to find a good name for the function. I really don't know if it makes a difference on the total time vol_to_surf takes to return (my guess is no)

Reply:
I have written a patch inverting this function to _kept_indices. playing around with %timeit I see no difference, which makes sense: it takes 1 microsecond to do a logical not on an array of the size of fsaverage on my machine; around 6 microseconds for an array ten times bigger. Since we are dealing with dozens of milliseconds in vol_to_surf, this is very small. should I push this commit anyway?